x-pack/plugins/observability/server/services/slo/historical_summary_client.ts

```diff
@@ -0,0 +1,302 @@
+      const errorBudget = toErrorBudget(initialErrorBudget, consumedErrorBudget);
+
+      return {
+        date: new Date(bucket.key_as_string),
+        errorBudget,
+        sliValue,
+        status: computeSummaryStatus(slo, sliValue, errorBudget),
+      };
+    });
+}
+
+function generateSearchQuery(slo: SLO, dateRange: DateRange): MsearchMultisearchBody {
+  const unit = toMomentUnitOfTime(slo.timeWindow.duration.unit);
+  const timeWindowDurationInDays = moment.duration(slo.timeWindow.duration.value, unit).asDays();
+
+  return {
+    size: 0,
+    query: {
+      bool: {
+        filter: [
+          { term: { 'slo.id': slo.id } },
+          { term: { 'slo.revision': slo.revision } },
+          {
+            range: {
+              '@timestamp': {
+                gte: dateRange.from.toISOString(),
+                lte: dateRange.to.toISOString(),
+              },
+            },
+          },
+        ],
+      },
+    },
+    aggs: {
+      daily: {
+        date_histogram: {
+          field: '@timestamp',
+          fixed_interval: '1d',
+          extended_bounds: {
+            min: dateRange.from.toISOString(),
+            max: 'now/d',
+          },
+        },
+        aggs: {
+          ...(occurrencesBudgetingMethodSchema.is(slo.budgetingMethod) && {
+            good: {
+              sum: {
+                field: 'slo.numerator',
+              },
+            },
+            total: {
+              sum: {
+                field: 'slo.denominator',
+              },
+            },
+          }),
+          ...(timeslicesBudgetingMethodSchema.is(slo.budgetingMethod) && {
+            good: {
+              sum: {
+                field: 'slo.isGoodSlice',
+              },
+            },
+            total: {
+              value_count: {
+                field: 'slo.isGoodSlice',
+              },
+            },
+          }),
+          cumulative_good: {
+            moving_fn: {
+              buckets_path: 'good',
+              window: timeWindowDurationInDays,
+              shift: 1,
+              script: 'MovingFunctions.sum(values)',
+            },
+          },
+          cumulative_total: {
+            moving_fn: {
+              buckets_path: 'total',
+              window: timeWindowDurationInDays,
+              shift: 1,
+              script: 'MovingFunctions.sum(values)',
+            },
+          },
+        },
+      },
+    },
+  };
+}
+
+function getDateRange(slo: SLO) {
+  const unit = toMomentUnitOfTime(slo.timeWindow.duration.unit);
+
+  if (rollingTimeWindowSchema.is(slo.timeWindow)) {
+    const now = moment();
+    return {
+      from: now
+        .clone()
+        .subtract(slo.timeWindow.duration.value * 2, unit)
+        .startOf('day')
+        .toDate(),
+      to: now.startOf('minute').toDate(),
+    };
+  }
+  if (calendarAlignedTimeWindowSchema.is(slo.timeWindow)) {
+    return toDateRange(slo.timeWindow);
+  }
+
+  assertNever(slo.timeWindow);
+}
```

Comment:
💬 For rolling we need to compute twice the duration in days, for calendar aligned, we use the toDateRange utility method